// @aws-cdk/aws-codepipeline/test/integ.cfn-template-from-repo.lit.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const cfn = require("@aws-cdk/aws-cloudformation");
const codecommit = require("@aws-cdk/aws-codecommit");
const cdk = require("@aws-cdk/cdk");
const codepipeline = require("../lib");
const app = new cdk.App();
const stack = new cdk.Stack(app, 'aws-cdk-codepipeline-cloudformation');
/// !show
const pipeline = new codepipeline.Pipeline(stack, 'Pipeline');
// Source stage: read from repository
const repo = new codecommit.Repository(stack, 'TemplateRepo', {
    repositoryName: 'template-repo'
});
const sourceStage = new codepipeline.Stage(pipeline, 'Source', { pipeline });
const source = new codecommit.PipelineSourceAction(stack, 'Source', {
    stage: sourceStage,
    repository: repo,
    outputArtifactName: 'SourceArtifact',
    pollForSourceChanges: true,
});
// Deployment stage: create and deploy changeset with manual approval
const prodStage = new codepipeline.Stage(pipeline, 'Deploy', { pipeline });
const stackName = 'OurStack';
const changeSetName = 'StagedChangeSet';
new cfn.PipelineCreateReplaceChangeSetAction(prodStage, 'PrepareChanges', {
    stage: prodStage,
    stackName,
    changeSetName,
    adminPermissions: true,
    templatePath: source.outputArtifact.atPath('template.yaml'),
});
new codepipeline.ManualApprovalAction(stack, 'ApproveChanges', {
    stage: prodStage,
});
new cfn.PipelineExecuteChangeSetAction(stack, 'ExecuteChanges', {
    stage: prodStage,
    stackName,
    changeSetName,
});
/// !hide
app.run();
