// @aws-cdk/aws-codepipeline/test/integ.pipeline-events.js
"use strict";
// Use pipeline as CloudWAtch event target
Object.defineProperty(exports, "__esModule", { value: true });
const codebuild = require("@aws-cdk/aws-codebuild");
const codecommit = require("@aws-cdk/aws-codecommit");
const sns = require("@aws-cdk/aws-sns");
const cdk = require("@aws-cdk/cdk");
const codepipeline = require("../lib");
const app = new cdk.App();
const stack = new cdk.Stack(app, 'aws-cdk-pipeline-event-target');
const pipeline = new codepipeline.Pipeline(stack, 'MyPipeline');
const sourceStage = new codepipeline.Stage(stack, 'Source', { pipeline });
const buildStage = new codepipeline.Stage(stack, 'Build', { pipeline });
const repository = new codecommit.Repository(stack, 'CodeCommitRepo', {
    repositoryName: 'foo'
});
const project = new codebuild.PipelineProject(stack, 'BuildProject');
const sourceAction = new codecommit.PipelineSourceAction(pipeline, 'CodeCommitSource', {
    stage: sourceStage,
    outputArtifactName: 'Source',
    repository,
    pollForSourceChanges: true,
});
new codebuild.PipelineBuildAction(stack, 'CodeBuildAction', {
    stage: buildStage,
    inputArtifact: sourceAction.outputArtifact,
    project
});
const topic = new sns.Topic(stack, 'MyTopic');
pipeline.onStateChange('OnPipelineStateChange').addTarget(topic, {
    textTemplate: 'Pipeline <pipeline> changed state to <state>',
    pathsMap: {
        pipeline: '$.detail.pipeline',
        state: '$.detail.state'
    }
});
sourceStage.onStateChange('OnSourceStateChange', topic);
sourceAction.onStateChange('OnActionStateChange', topic).addEventPattern({
    detail: { state: ['STARTED'] }
});
app.run();
