// @aws-cdk/aws-codepipeline/test/integ.pipeline-cfn.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const cfn = require("@aws-cdk/aws-cloudformation");
const aws_iam_1 = require("@aws-cdk/aws-iam");
const aws_iam_2 = require("@aws-cdk/aws-iam");
const s3 = require("@aws-cdk/aws-s3");
const cdk = require("@aws-cdk/cdk");
const codepipeline = require("../lib");
const app = new cdk.App();
const stack = new cdk.Stack(app, 'aws-cdk-codepipeline-cloudformation');
const pipeline = new codepipeline.Pipeline(stack, 'Pipeline');
const sourceStage = new codepipeline.Stage(pipeline, 'Source', { pipeline });
const bucket = new s3.Bucket(stack, 'PipelineBucket', {
    versioned: true,
    removalPolicy: cdk.RemovalPolicy.Destroy,
});
const source = new s3.PipelineSourceAction(stack, 'Source', {
    stage: sourceStage,
    outputArtifactName: 'SourceArtifact',
    bucket,
    bucketKey: 'key',
});
const cfnStage = new codepipeline.Stage(stack, 'CFN', { pipeline });
const changeSetName = "ChangeSetIntegTest";
const stackName = "IntegTest-TestActionStack";
const role = new aws_iam_1.Role(stack, 'CfnChangeSetRole', {
    assumedBy: new aws_iam_2.ServicePrincipal('cloudformation.amazonaws.com'),
});
new cfn.PipelineCreateReplaceChangeSetAction(stack, 'DeployCFN', {
    stage: cfnStage,
    changeSetName,
    stackName,
    deploymentRole: role,
    templatePath: source.outputArtifact.atPath('test.yaml'),
    adminPermissions: false,
    parameterOverrides: {
        BucketName: source.outputArtifact.bucketName,
        ObjectKey: source.outputArtifact.objectKey,
        Url: source.outputArtifact.url,
        OtherParam: source.outputArtifact.getParam('params.json', 'OtherParam'),
    },
});
app.run();
