// @aws-cdk/aws-codebuild/test/integ.project-events.js
#!/usr/bin/env node
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const codecommit = require("@aws-cdk/aws-codecommit");
const sns = require("@aws-cdk/aws-sns");
const sqs = require("@aws-cdk/aws-sqs");
const cdk = require("@aws-cdk/cdk");
const lib_1 = require("../lib");
const app = new cdk.App();
const stack = new cdk.Stack(app, 'aws-cdk-codebuild-events');
const repo = new codecommit.Repository(stack, 'MyRepo', { repositoryName: 'aws-cdk-codebuild-events' });
const project = new lib_1.Project(stack, 'MyProject', {
    source: new lib_1.CodeCommitSource({ repository: repo }),
});
const queue = new sqs.Queue(stack, 'MyQueue');
const topic = new sns.Topic(stack, 'MyTopic');
topic.subscribeQueue(queue);
// this will send an email with the JSON event for every state change of this
// build project.
project.onStateChange('StateChange', topic);
// this will send an email with the message "Build phase changed to <phase>".
// The phase will be extracted from the "completed-phase" field of the event
// details.
project.onPhaseChange('PhaseChange').addTarget(topic, {
    textTemplate: `Build phase changed to <phase>`,
    pathsMap: {
        phase: '$.detail.completed-phase'
    }
});
// trigger a build when a commit is pushed to the repo
const onCommitRule = repo.onCommit('OnCommit', project, 'master');
onCommitRule.addTarget(topic, {
    textTemplate: 'A commit was pushed to the repository <repo> on branch <branch>',
    pathsMap: {
        branch: '$.detail.referenceName',
        repo: '$.detail.repositoryName'
    }
});
app.run();
