// @aws-cdk/cdk/test/cloudformation/test.stack.js
"use strict";
const cxapi = require("@aws-cdk/cx-api");
const lib_1 = require("../../lib");
class StackWithPostProcessor extends lib_1.Stack {
    // ...
    toCloudFormation() {
        const template = super.toCloudFormation();
        // manipulate template (e.g. rename "Key" to "key")
        template.Resources.myResource.Properties.Environment.key =
            template.Resources.myResource.Properties.Environment.Key;
        delete template.Resources.myResource.Properties.Environment.Key;
        return template;
    }
}
module.exports = {
    'a stack can be serialized into a CloudFormation template, initially it\'s empty'(test) {
        const stack = new lib_1.Stack();
        test.deepEqual(stack.toCloudFormation(), {});
        test.done();
    },
    'stack objects have some template-level propeties, such as Description, Version, Transform'(test) {
        const stack = new lib_1.Stack();
        stack.templateOptions.templateFormatVersion = 'MyTemplateVersion';
        stack.templateOptions.description = 'This is my description';
        stack.templateOptions.transform = 'SAMy';
        test.deepEqual(stack.toCloudFormation(), {
            Description: 'This is my description',
            AWSTemplateFormatVersion: 'MyTemplateVersion',
            Transform: 'SAMy'
        });
        test.done();
    },
    'Stack.find(c) can be used to find the stack from any point in the tree'(test) {
        const stack = new lib_1.Stack();
        const level1 = new lib_1.Construct(stack, 'level1');
        const level2 = new lib_1.Construct(level1, 'level2');
        const level3 = new lib_1.Construct(level2, 'level3');
        const res1 = new lib_1.Resource(level1, 'childoflevel1', { type: 'MyResourceType1' });
        const res2 = new lib_1.Resource(level3, 'childoflevel3', { type: 'MyResourceType2' });
        test.equal(lib_1.Stack.find(res1), stack);
        test.equal(lib_1.Stack.find(res2), stack);
        test.equal(lib_1.Stack.find(level2), stack);
        const root = new lib_1.Root();
        const child = new lib_1.Construct(root, 'child');
        test.throws(() => lib_1.Stack.find(child));
        test.throws(() => lib_1.Stack.find(root));
        test.done();
    },
    'Stack.isStack indicates that a construct is a stack'(test) {
        const stack = new lib_1.Stack();
        const c = new lib_1.Construct(stack, 'Construct');
        test.ok(lib_1.Stack.isStack(stack));
        test.ok(!lib_1.Stack.isStack(c));
        test.done();
    },
    'stack.id is not included in the logical identities of resources within it'(test) {
        const stack = new lib_1.Stack(undefined, 'MyStack');
        new lib_1.Resource(stack, 'MyResource', { type: 'MyResourceType' });
        test.deepEqual(stack.toCloudFormation(), { Resources: { MyResource: { Type: 'MyResourceType' } } });
        test.done();
    },
    'stack.templateOptions can be used to set template-level options'(test) {
        const stack = new lib_1.Stack();
        stack.templateOptions.description = 'StackDescription';
        stack.templateOptions.templateFormatVersion = 'TemplateVersion';
        stack.templateOptions.transform = 'Transform';
        stack.templateOptions.metadata = {
            MetadataKey: 'MetadataValue'
        };
        test.deepEqual(stack.toCloudFormation(), {
            Description: 'StackDescription',
            Transform: 'Transform',
            AWSTemplateFormatVersion: 'TemplateVersion',
            Metadata: { MetadataKey: 'MetadataValue' }
        });
        test.done();
    },
    // This approach will only apply to TypeScript code, but at least it's a temporary
    // workaround for people running into issues caused by SDK-3003.
    // We should come up with a proper solution that involved jsii callbacks (when they exist)
    // so this can be implemented by jsii languages as well.
    'Overriding `Stack.toCloudFormation` allows arbitrary post-processing of the generated template during synthesis'(test) {
        const stack = new StackWithPostProcessor();
        new lib_1.Resource(stack, 'myResource', {
            type: 'AWS::MyResource',
            properties: {
                MyProp1: 'hello',
                MyProp2: 'howdy',
                Environment: {
                    Key: 'value'
                }
            }
        });
        test.deepEqual(stack.toCloudFormation(), { Resources: { myResource: { Type: 'AWS::MyResource',
                    Properties: { MyProp1: 'hello',
                        MyProp2: 'howdy',
                        Environment: { key: 'value' } } } } });
        test.done();
    },
    'Construct.findResource(logicalId) can be used to retrieve a resource by its path'(test) {
        const stack = new lib_1.Stack();
        test.ok(!stack.node.tryFindChild('foo'), 'empty stack');
        const r1 = new lib_1.Resource(stack, 'Hello', { type: 'MyResource' });
        test.equal(stack.findResource(r1.stackPath), r1, 'look up top-level');
        const child = new lib_1.Construct(stack, 'Child');
        const r2 = new lib_1.Resource(child, 'Hello', { type: 'MyResource' });
        test.equal(stack.findResource(r2.stackPath), r2, 'look up child');
        test.done();
    },
    'Stack.findResource will fail if the element is not a resource'(test) {
        const stack = new lib_1.Stack();
        const p = new lib_1.Parameter(stack, 'MyParam', { type: 'String' });
        test.throws(() => stack.findResource(p.node.path));
        test.done();
    },
    'Stack.getByPath can be used to find any CloudFormation element (Parameter, Output, etc)'(test) {
        const stack = new lib_1.Stack();
        const p = new lib_1.Parameter(stack, 'MyParam', { type: 'String' });
        const o = new lib_1.Output(stack, 'MyOutput');
        const c = new lib_1.Condition(stack, 'MyCondition');
        test.equal(stack.node.findChild(p.node.path), p);
        test.equal(stack.node.findChild(o.node.path), o);
        test.equal(stack.node.findChild(c.node.path), c);
        test.done();
    },
    'Stack names can have hyphens in them'(test) {
        const root = new lib_1.App();
        new lib_1.Stack(root, 'Hello-World');
        // Did not throw
        test.done();
    },
    'Include should support non-hash top-level template elements like "Description"'(test) {
        const stack = new lib_1.Stack();
        const template = {
            Description: 'hello, world'
        };
        new lib_1.Include(stack, 'Include', { template });
        const output = stack.toCloudFormation();
        test.equal(typeof output.Description, 'string');
        test.done();
    },
    'Pseudo values attached to one stack can be referenced in another stack'(test) {
        // GIVEN
        const app = new lib_1.App();
        const stack1 = new lib_1.Stack(app, 'Stack1');
        const account1 = new lib_1.Aws(stack1).accountId;
        const stack2 = new lib_1.Stack(app, 'Stack2');
        // WHEN - used in another stack
        new lib_1.Parameter(stack2, 'SomeParameter', { type: 'String', default: account1 });
        // THEN
        // Need to do this manually now, since we're in testing mode. In a normal CDK app,
        // this happens as part of app.run().
        app.node.prepareTree();
        test.deepEqual(stack1.toCloudFormation(), {
            Outputs: {
                ExportsOutputRefAWSAccountIdAD568057: {
                    Value: { Ref: 'AWS::AccountId' },
                    Export: { Name: 'Stack1:ExportsOutputRefAWSAccountIdAD568057' }
                }
            }
        });
        test.deepEqual(stack2.toCloudFormation(), {
            Parameters: {
                SomeParameter: {
                    Type: 'String',
                    Default: { 'Fn::ImportValue': 'Stack1:ExportsOutputRefAWSAccountIdAD568057' }
                }
            }
        });
        test.done();
    },
    'cross-stack references in lazy tokens work'(test) {
        // GIVEN
        const app = new lib_1.App();
        const stack1 = new lib_1.Stack(app, 'Stack1');
        const account1 = new lib_1.Aws(stack1).accountId;
        const stack2 = new lib_1.Stack(app, 'Stack2');
        // WHEN - used in another stack
        new lib_1.Parameter(stack2, 'SomeParameter', { type: 'String', default: new lib_1.Token(() => account1) });
        app.node.prepareTree();
        // THEN
        test.deepEqual(stack1.toCloudFormation(), {
            Outputs: {
                ExportsOutputRefAWSAccountIdAD568057: {
                    Value: { Ref: 'AWS::AccountId' },
                    Export: { Name: 'Stack1:ExportsOutputRefAWSAccountIdAD568057' }
                }
            }
        });
        test.deepEqual(stack2.toCloudFormation(), {
            Parameters: {
                SomeParameter: {
                    Type: 'String',
                    Default: { 'Fn::ImportValue': 'Stack1:ExportsOutputRefAWSAccountIdAD568057' }
                }
            }
        });
        test.done();
    },
    'cross-stack references in strings work'(test) {
        // GIVEN
        const app = new lib_1.App();
        const stack1 = new lib_1.Stack(app, 'Stack1');
        const account1 = new lib_1.Aws(stack1).accountId;
        const stack2 = new lib_1.Stack(app, 'Stack2');
        // WHEN - used in another stack
        new lib_1.Parameter(stack2, 'SomeParameter', { type: 'String', default: `TheAccountIs${account1}` });
        app.node.prepareTree();
        // THEN
        test.deepEqual(stack2.toCloudFormation(), {
            Parameters: {
                SomeParameter: {
                    Type: 'String',
                    Default: { 'Fn::Join': ['', ['TheAccountIs', { 'Fn::ImportValue': 'Stack1:ExportsOutputRefAWSAccountIdAD568057' }]] }
                }
            }
        });
        test.done();
    },
    'cannot create cyclic reference between stacks'(test) {
        // GIVEN
        const app = new lib_1.App();
        const stack1 = new lib_1.Stack(app, 'Stack1');
        const account1 = new lib_1.Aws(stack1).accountId;
        const stack2 = new lib_1.Stack(app, 'Stack2');
        const account2 = new lib_1.Aws(stack2).accountId;
        // WHEN
        new lib_1.Parameter(stack2, 'SomeParameter', { type: 'String', default: account1 });
        new lib_1.Parameter(stack1, 'SomeParameter', { type: 'String', default: account2 });
        test.throws(() => {
            app.node.prepareTree();
        }, /Adding this dependency would create a cyclic reference/);
        test.done();
    },
    'stacks know about their dependencies'(test) {
        // GIVEN
        const app = new lib_1.App();
        const stack1 = new lib_1.Stack(app, 'Stack1');
        const account1 = new lib_1.Aws(stack1).accountId;
        const stack2 = new lib_1.Stack(app, 'Stack2');
        // WHEN
        new lib_1.Parameter(stack2, 'SomeParameter', { type: 'String', default: account1 });
        app.node.prepareTree();
        // THEN
        test.deepEqual(stack2.dependencies().map(s => s.node.id), ['Stack1']);
        test.done();
    },
    'cannot create references to stacks in other regions/accounts'(test) {
        // GIVEN
        const app = new lib_1.App();
        const stack1 = new lib_1.Stack(app, 'Stack1', { env: { account: '123456789012', region: 'es-norst-1' } });
        const account1 = new lib_1.Aws(stack1).accountId;
        const stack2 = new lib_1.Stack(app, 'Stack2', { env: { account: '123456789012', region: 'es-norst-2' } });
        // WHEN
        new lib_1.Parameter(stack2, 'SomeParameter', { type: 'String', default: account1 });
        test.throws(() => {
            app.node.prepareTree();
        }, /Can only reference cross stacks in the same region and account/);
        test.done();
    },
    'stack with region supplied via props returns literal value'(test) {
        // GIVEN
        const app = new lib_1.App();
        const stack = new lib_1.Stack(app, 'Stack1', { env: { account: '123456789012', region: 'es-norst-1' } });
        // THEN
        test.equal(stack.node.resolve(stack.region), 'es-norst-1');
        test.done();
    },
    'stack with region supplied via context returns symbolic value'(test) {
        // GIVEN
        const app = new lib_1.App();
        app.node.setContext(cxapi.DEFAULT_REGION_CONTEXT_KEY, 'es-norst-1');
        const stack = new lib_1.Stack(app, 'Stack1');
        // THEN
        test.deepEqual(stack.node.resolve(stack.region), { Ref: 'AWS::Region' });
        test.done();
    },
    'overrideLogicalId(id) can be used to override the logical ID of a resource'(test) {
        // GIVEN
        const stack = new lib_1.Stack();
        const bonjour = new lib_1.Resource(stack, 'BonjourResource', { type: 'Resource::Type' });
        // { Ref } and { GetAtt }
        new lib_1.Resource(stack, 'RefToBonjour', { type: 'Other::Resource', properties: {
                RefToBonjour: bonjour.ref.toString(),
                GetAttBonjour: bonjour.getAtt('TheAtt').toString()
            } });
        bonjour.overrideLogicalId('BOOM');
        // THEN
        test.deepEqual(stack.toCloudFormation(), { Resources: { BOOM: { Type: 'Resource::Type' },
                RefToBonjour: { Type: 'Other::Resource',
                    Properties: { RefToBonjour: { Ref: 'BOOM' },
                        GetAttBonjour: { 'Fn::GetAtt': ['BOOM', 'TheAtt'] } } } } });
        test.done();
    }
};
